--- a/osdep/threads-posix.h
+++ b/osdep/threads-posix.h
@@ -217,7 +217,9 @@ static inline void mp_thread_set_name(const char *name)
 {
-#if HAVE_GLIBC_THREAD_NAME
+#if HAVE_GLIBC_THREAD_NAME && !defined(__APPLE__)
     if (pthread_setname_np(pthread_self(), name) == ERANGE) {
         char tname[16] = {0}; // glibc-checked kernel limit
         strncpy(tname, name, sizeof(tname) - 1);
         pthread_setname_np(pthread_self(), tname);
     }
+#elif (HAVE_GLIBC_THREAD_NAME && defined(__APPLE__)) || HAVE_MAC_THREAD_NAME
+    pthread_setname_np(name);
 #elif HAVE_BSD_THREAD_NAME
     pthread_set_name_np(pthread_self(), name);
 #elif HAVE_MAC_THREAD_NAME
